### PR TITLE
 Add metrics api for poseidon

### PR DESCRIPTION
**Pull request by jiaxuanzhou**
fix： #94 
1, we will collect metrics like scheduling latency/binding latency ..
2, port for metrics will be configurable, default as port  9292 or can we multiplex other ports?
3, need to provide capability like health check(like :/healthz) for poseidon? 
@m1093782566

### CODE REVIEW

**Comment by m1093782566** [comment]
@jiaxuanzhou 

Thanks for your PR. You need to

* Sign the CLA.

* Add your vendors to the `Gopkg.lock`.

**Comment by m1093782566** [comment]
@jiaxuanzhou 

Is that ready for reviewing?

**Comment by jiaxuanzhou** [comment]
@m1093782566 i will submit the code soon

**Comment by m1093782566** [comment]
@jiaxuanzhou Please remove the WIP once this PR is ready for reviewing.

**Comment by jiaxuanzhou** [comment]
@m1093782566 plse help review

**Comment by jiaxuanzhou** [comment]
@m1093782566 please hold, will add the service of metrics port later

**Comment by jiaxuanzhou** [comment]
@m1093782566 please help review, healthz funcs  will multiplex the port of metrics.

**Comment by m1093782566** [comment]
@jiaxuanzhou 

Can you please clean up the commits history? e.g. squeeze some similar changes but keep the commit for vendor.

**Comment by jiaxuanzhou** [comment]
@m1093782566 sure

**Comment by jiaxuanzhou** [comment]
@m1093782566 rebased the commits

File: pkg/metrics/BUILD [hunk 1]
```diff
@@ -5,9 +5,5 @@ go_library(
     srcs = ["metrics.go"],
     importpath = "github.com/kubernetes-sigs/poseidon/pkg/metrics",
     visibility = ["//visibility:public"],
-<<<<<<< HEAD
```
Comment (m1093782566) [inline pkg/metrics/BUILD]: Please fix the git error.

Comment (jiaxuanzhou) [reply pkg/metrics/BUILD]: fixed

File: pkg/poseidonhttp/poseidonhttp.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+		Addr:    addr,
+		Handler: metricsMux,
+	}
+	go glog.Fatal(server.ListenAndServe())
+}
+
+// HandleMetricsHealth registers metrics and health handlers.
+func HandleMetricsHealth(mux *http.ServeMux, fc firmament.FirmamentSchedulerClient) {
+	mux.Handle(pathMetrics, promhttp.Handler())
+	// TODO(jiaxuanzhou): add healthz handler
```
Comment (m1093782566) [inline pkg/poseidonhttp/poseidonhttp.go]: What's the problem that we can't support healthz endpoint now?

Comment (jiaxuanzhou) [reply pkg/poseidonhttp/poseidonhttp.go]: i will implement it in another pr

File: pkg/metrics/metrics.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+		})
+)
+
+var registerMetrics sync.Once
+
+// Register all metrics.
+func Register() {
+	// Register the metrics.
+	registerMetrics.Do(func() {
+		prometheus.MustRegister(SchedulingSubmitmLatency) //
```
Comment (m1093782566) [inline pkg/metrics/metrics.go]: nit: remove the empty comment line.

File: pkg/config/config.go [hunk 4]
```diff
@@ -137,6 +143,7 @@ func ReadFromCommandLineFlags() {
 		"The path to the config file (i.e poseidon_cfg) without filename or extension, supported extensions/formats are Yaml, Json")
 	flag.BoolVar(&config.EnablePprof, "enablePprof", false, "Enable runtime profiling data via HTTP server. Address is at client URL + \"/debug/pprof/\"")
 	flag.StringVar(&config.PprofAddress, "pprofAddress", "0.0.0.0:8989", "Address on which to collect runtime profiling data,default to set for all interfaces ")
+	pflag.StringVar(&config.Port, "port", "9292", "The port that the poseidon's http service runs on,default to set for all interfaces")
```
Comment (m1093782566) [inline pkg/config/config.go]: I would prefer MetricsBindAddress than port.

Comment (m1093782566) [reply pkg/config/config.go]: Plus, I would suggest default metrics(and probably healthz in the future) reuse the same port - 8989.

Comment (jiaxuanzhou) [reply pkg/config/config.go]: agree with u, commit submitted

File: pkg/config/config.go [hunk 5]
```diff
@@ -137,6 +148,7 @@ func ReadFromCommandLineFlags() {
 		"The path to the config file (i.e poseidon_cfg) without filename or extension, supported extensions/formats are Yaml, Json")
 	flag.BoolVar(&config.EnablePprof, "enablePprof", false, "Enable runtime profiling data via HTTP server. Address is at client URL + \"/debug/pprof/\"")
 	flag.StringVar(&config.PprofAddress, "pprofAddress", "0.0.0.0:8989", "Address on which to collect runtime profiling data,default to set for all interfaces ")
+	pflag.StringVar(&config.MetricsBindAddress, "metricsBindAddress", "0.0.0.0:8989", "The port that the poseidon's http service runs on,default to set for all interfaces")
```
Comment (m1093782566) [inline pkg/config/config.go]: parameter message does not match the parameter. It should be something like,

Address on which to serve poseidon runtime metrics.

**Comment by m1093782566** [comment]
/approve except some nits.

@shivramsrivastava any chance to take a look at this PR?

File: pkg/config/config.go (hunk 5 above)
Comment (jiaxuanzhou) [reply pkg/config/config.go]: eeem,  good point

**Pull request merged**